### PR TITLE
Fixed the rendering of the dynamic group associable model list tables.

### LINKED ISSUE #6661: Rendering a table of a dynamic_group_associable model from list data fails
### Environment
* Nautobot version (Docker tag too if applicable):
* Python version:
* Database platform, version:
* Middleware(s):

### Steps to Reproduce
1. Go to https://next.demo.nautobot.com/wireless/wireless-networks/import/
2. Try to import any csv data. Example:
```csv
name,ssid,mode,authentication
testwirelessnetwork,testssid,Central,Open
```

### Expected Behavior

Bulk import performed without any errors

### Observed Behavior

The bulk import is performed but the generation of the table for the returned response raises an exception:

```
nautobot-1  | --- Logging error ---
nautobot-1  | Traceback (most recent call last):
nautobot-1  |   File "/usr/local/lib/python3.11/site-packages/django/core/handlers/exception.py", line 55, in inner
nautobot-1  |     response = get_response(request)
nautobot-1  |                ^^^^^^^^^^^^^^^^^^^^^
nautobot-1  |   File "/usr/local/lib/python3.11/site-packages/django/core/handlers/base.py", line 220, in _get_response
nautobot-1  |     response = response.render()
nautobot-1  |                ^^^^^^^^^^^^^^^^^
nautobot-1  |   File "/usr/local/lib/python3.11/site-packages/django/template/response.py", line 114, in render
nautobot-1  |     self.content = self.rendered_content
nautobot-1  |                    ^^^^^^^^^^^^^^^^^^^^^
nautobot-1  |   File "/usr/local/lib/python3.11/site-packages/rest_framework/response.py", line 74, in rendered_content
nautobot-1  |     ret = renderer.render(self.data, accepted_media_type, context)
nautobot-1  |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
nautobot-1  |   File "/usr/local/lib/python3.11/site-packages/nautobot/core/views/renderers.py", line 370, in render
nautobot-1  |     return super().render(data, accepted_media_type=accepted_media_type, renderer_context=renderer_context)
nautobot-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
nautobot-1  |   File "/usr/local/lib/python3.11/site-packages/rest_framework/renderers.py", line 727, in render
nautobot-1  |     context = self.get_context(data, accepted_media_type, renderer_context)
nautobot-1  |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
nautobot-1  |   File "/usr/local/lib/python3.11/site-packages/nautobot/core/views/renderers.py", line 292, in get_context
nautobot-1  |     "table_config_form": TableConfigForm(table=table) if table else None,
nautobot-1  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
nautobot-1  |   File "/usr/local/lib/python3.11/site-packages/nautobot/core/forms/forms.py", line 282, in __init__
nautobot-1  |     self.fields["columns"].choices = table.configurable_columns
nautobot-1  |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
nautobot-1  |   File "/usr/local/lib/python3.11/site-packages/nautobot/core/tables.py", line 275, in configurable_columns
nautobot-1  |     selected_columns = [
nautobot-1  |                        ^
nautobot-1  |   File "/usr/local/lib/python3.11/site-packages/nautobot/core/tables.py", line 276, in <listcomp>
nautobot-1  |     (name, self.columns[name].verbose_name) for name in self.sequence if name not in ["pk", "actions"]
nautobot-1  |            ~~~~~~~~~~~~^^^^^^
nautobot-1  |   File "/usr/local/lib/python3.11/site-packages/django_tables2/columns/base.py", line 848, in __getitem__
nautobot-1  |     raise KeyError(
nautobot-1  | KeyError: "Column with name 'dynamic_group_count' does not exist; choices are: ['pk', 'name', 'ssid', 'mode', 'authentication', 'enabled', 'hidden', 'secret', 'description', 'tags', 'actions']"
```

Even though this view has been deprecated in core, some apps still use the legacy bulk import view. One example is golden config. This is either a regression from #6626 or a bug in our `BaseTable` implementation that was identified by the changes in that PR. When we exclude table columns they are still appearing in the table's `sequence` property. I think core is accessing the table's `sequence` property incorrectly when `columns.names()` should be used.

```
> /usr/local/lib/python3.11/site-packages/nautobot/core/tables.py(317)configurable_columns()
-> selected_columns = [
(Pdb) self.columns.names()
['pk', 'name', 'platform', 'description', 'regex', 'replace']
(Pdb) self.base_columns
OrderedDict([('name', <django_tables2.columns.linkcolumn.LinkColumn object at 0x7f65031f4b10>), ('platform', <django_tables2.columns.base.Column object at 0x7f65031f5990>), ('description', <django_tables2.columns.base.Column object at 0x7f65031f5a90>), ('regex', <django_tables2.columns.base.Column object at 0x7f65031f5bd0>), ('replace', <django_tables2.columns.base.Column object at 0x7f65031f5b50>), ('dynamic_group_count', <nautobot.core.tables.LinkedCountColumn object at 0x7f64ff37e1d0>), ('pk', <nautobot.core.tables.ToggleColumn object at 0x7f65031f5350>)])
(Pdb) self.sequence
['pk', 'name', 'platform', 'description', 'regex', 'replace', 'dynamic_group_count']
(Pdb) self.exclude
['dynamic_group_count']
(Pdb) 
```

### PR DESCRIPTION
**Pull request by HanlinMiao**
# Closes #6661 
# What's Changed

Fixed the rendering of the dynamic group associable model list tables.
# Screenshots

Legacy import view works now.
Screenshot 2024-12-16 at 5 02 19 PM
Wireless Network List view with `dynamic_group_count` column
Screenshot 2024-12-16 at 5 15 54 PM

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/core/tables.py [hunk 1]
```diff
@@ -145,13 +145,18 @@ def __init__(
...
             for name, column in self.base_columns.items():
                 if name in columns and name not in self.exclude:
                     self.columns.show(name)
                 else:
                     self.columns.hide(name)
             self.sequence = [c for c in columns if c in self.base_columns]
+        else:
+            for c in self.base_columns:
+                if c not in self.exclude:
+                    self.sequence.append(c)
```
Comment (glennmatthews) [inline nautobot/core/tables.py:159]: This block feels "off" to me as I look at it. `self.sequence` is the *order* columns are listed in, which should be independent of which columns are visible or hidden. The `if columns` case makes some sense to me, as it is using the user preferences to both 1) reorder the sequence and 2) show/hide columns as per the user preferences, but it seems wrong to me to recalculate `self.sequence` in the `else` case. 

I think @gsnider2195 analysis in #6661 is correct in stating that:

> I think core is accessing the table's sequence property incorrectly when columns.names() should be used.

as `Table.columns` *does* respect `self.exclude` whereas `Table.sequence` does not and should not. I think it's likely that both `configurable_columns()` and `visible_columns()` are implemented incorrectly as such.

Comment (glennmatthews) [reply nautobot/core/tables.py:159]: I think we should also try to understand how/why `dynamic_group_count` is getting incorrectly added to `self.sequence`, though.

**Review by glennmatthews — COMMENTED** [review]
Looks more reasonable to me now, thanks. Do we have any testing for these APIs, and if not, can you add some?

File: nautobot/core/tests/test_tables.py [hunk 2]
```diff
@@ -63,3 +67,55 @@ def test_tree_model_table_orderable(self):
...
+        # Wireless Network Table
+        for name, column in wn_table.columns.items():
+            expected_visible_columns = [
+                name for name, column in wn_table.columns.items() if column.visible and name not in wn_table.exclude
+            ]
+            expected_configurable_columns = [
+                (name, column.verbose_name)
+                for name, column in wn_table.columns.items()
+                if name not in ["pk", "actions"]
+            ]
```
Comment (glennmatthews) [inline nautobot/core/tests/test_tables.py:89]: Feels a bit cheaty to just copy/paste the same logic into the test case. :-)

Can we instead actually mess with the tables a bit (`LocationTable(Location.objects.all(), exclude=[...])`, `location_table.columns["foo"].hide()`, etc.) and assert some specific lists of expected values from these APIs as a consequence?

Comment (HanlinMiao) [reply nautobot/core/tests/test_tables.py:89]: That's what I did originally, my concern is that we need to change the test case every time we add/remove columns from the table. But I can change it.

**Review by glennmatthews — APPROVED** [review]

**Comment by glennmatthews** [comment]
Does this need to be reworked further to address https://github.com/nautobot/nautobot/issues/6661#issuecomment-2550112223?

File: nautobot/core/tables.py [hunk 3]
```diff
@@ -64,6 +64,9 @@ def __init__(
                 returns new data. Runs after all of the queryset auto-optimization performed by this class.
                 Used for example in IPAM views to inject "fake" records for "available" Prefixes, IPAddresses, or VLANs.
             **kwargs (dict, optional): Passed through to django_tables2.Table
+        .. warning::
+            Do not modify/set the `base_columns` attribute after BaseTable class is instantiated.
+            Do not modify/set the `base_columns` attribute after calling super().__init__() of BaseTable class.
```
Comment (glennmatthews) [inline nautobot/core/tables.py:69]: May need to adjust this a bit to make it render well in mkdocs (https://docs-landing-page--6670.org.readthedocs.build/projects/core/en/6670/code-reference/nautobot/apps/tables/#nautobot.apps.tables.BaseTable):

image

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**